### PR TITLE
Tn class eval

### PR DESCRIPTION
**Pull request by yzhang123**

### CODE REVIEW

File: nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py [hunk 1]
```diff
@@ -158,20 +158,24 @@ def _infer(self, sents: List[List[str]], inst_directions: List[str]):
...
+        # Extract all_tag_preds for words
         all_tag_preds = []
         batch_size, max_len = encodings['input_ids'].size()
         for ix in range(batch_size):
-            raw_tag_preds = [constants.ALL_TAG_LABELS[p] for p in pred_indexes[ix][1:]]
+            raw_tag_preds = [
+                constants.ALL_TAG_LABELS[p] for p in pred_indexes[ix][2:]
+            ]  # remove first special token and task prefix token
             tag_preds, previous_word_idx = [], None
             word_ids = encodings.word_ids(batch_index=ix)
```
Comment (laituan245) [inline nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py:169]: Just set word_ids = encodings.word_ids(batch_index=ix)[2:] so you won't need the 3 lines from 171 to 173?

**Review by laituan245 — COMMENTED** [review]
These are my initial comments.

**Comment by laituan245** [comment]
In addition, in the [Google paper](https://research.fb.com/wp-content/uploads/2019/03/Neural-Models-of-Text-Normalization-for-Speech-Applications.pdf), when they evaluate the semiotic class accuracies, I believe they assume that the inputs are pre-segmented and the class-based scores in Table 8 are just for the decoder only. *

If that is the case, then I think it is cleaner to just create a new script for evaluating the class-based accuracies of the decoder like I sent you previously on Slack.

*It is just my guess since it is not clear from the descriptions how they did it.

In Table 5, they did not mention anything about semiotic class accuracies.  Notice that with ground-truth pre-segmented inputs, the token accuracy is 99.84%. This number 99.84% appears in Table 3, Table 4, and Table 7 (which all have semiotic class accuracies).

File: nemo/collections/nlp/data/text_normalization/test_dataset.py [hunk 2]
```diff
@@ -43,41 +44,95 @@ def __init__(self, input_file: str, mode: str, lang: str, keep_puncts: bool = Fa
...
                 if not s_word in constants.SPECIAL_WORDS:
                     processed_s_words.append(s_word)
                 processed_w_words.append(w_word)
+                processed_classes.append(cls)
+                processed_w_span_starts.append(w_word_idx)
+                processed_s_span_starts.append(s_word_idx)
+                processed_nb_spans += 1
+                num_words = len(basic_tokenize(w_word, lang=self.lang))
+                w_word_idx += num_words
+                s_word_idx += num_words
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/data/text_normalization/test_dataset.py:96]: good catch, thanks

File: nemo/collections/nlp/data/text_normalization/test_dataset.py [hunk 3]
```diff
@@ -127,3 +182,78 @@ def compute_sent_accuracy(preds: List[str], targets: List[str], inst_directions:
...
+            targets: List of lists where inner list contains target strings grouped by class boundary
+            tag_preds: List of lists where inner list contains predicted tags for each of the input words
+            inst_directions: A list of str where each str indicates the direction of the corresponding instance (i.e., INST_BACKWARD or INST_FORWARD).
+            output_spans: A list of lists where each inner list contains the decoded spans for the corresponding input sentence
+            classes: A list of lists where inner list contains the class for each semiotic token in input sentence
+            nb_spans: A list that contains the number of tokens in the input
+            span_starts: A list of lists where inner list contains the start word index of the current token
+            span_ends: A list of lists where inner list contains the end word index of the current token
+            lang: Language
+        Return: the sentence accuracy score
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/data/text_normalization/test_dataset.py:213]: good catch thanks

File: nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py (hunk 1 above)
Comment (yzhang123) [reply nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py:169]: thanks will update

**Review by ekmb — APPROVED** [review]

**Pull request merged**